rockcraft/plugins/jlink_plugin.py

```diff
@@ -0,0 +1,112 @@
+from overrides import override
+
+
+class JLinkPluginProperties(PluginProperties, frozen=True):
+    """The part properties used by the JLink plugin."""
+
+    plugin: Literal["jlink"] = "jlink"
+    jlink_java_version: int = 21
+    jlink_jars: list[str] = []
+
+
+class JLinkPlugin(Plugin):
+    """Create a Java Runtime using JLink."""
+
+    properties_class = JLinkPluginProperties
+
+    @override
+    def get_build_packages(self) -> set[str]:
+        options = cast(JLinkPluginProperties, self._options)
+        return {f"openjdk-{options.jlink_java_version}-jdk"}
+
+    @override
+    def get_build_environment(self) -> dict[str, str]:
+        return {}
+
+    @override
+    def get_build_snaps(self) -> set[str]:
+        return set()
+
+    @override
+    def get_build_commands(self) -> list[str]:
+        """Return a list of commands to run during the build step."""
+        options = cast(JLinkPluginProperties, self._options)
+
+        commands = []
+
+        if len(options.jlink_jars) > 0:
+            jars = " ".join(["${CRAFT_STAGE}/" + x for x in options.jlink_jars])
+            commands.append(f"PROCESS_JARS={jars}")
+        else:
+            commands.append("PROCESS_JARS=$(find ${CRAFT_STAGE} -type f -name *.jar)")
+
+        # create temp folder
+        commands.append("mkdir -p ${CRAFT_PART_BUILD}/tmp")
+        # extract jar files into temp folder
+        commands.append(
+            "(cd ${CRAFT_PART_BUILD}/tmp && for jar in ${PROCESS_JARS}; do jar xvf ${jar}; done;)"
+        )
+        commands.append("CPATH=$(find ${CRAFT_PART_BUILD}/tmp -type f -name *.jar)")
+        commands.append("CPATH=$(echo ${CPATH}:. | sed s'/[[:space:]]/:/'g)")
+        commands.append("echo ${CPATH}")
+        commands.append(
+            """if [ "x${PROCESS_JARS}" != "x" ]; then
+                deps=$(jdeps --class-path=${CPATH} -q --recursive  --ignore-missing-deps \
+                    --print-module-deps --multi-release """
+            + str(options.jlink_java_version)
+            + """ ${PROCESS_JARS}); else deps=java.base; fi
+            """
+        )
+        commands.append(
+            "INSTALL_ROOT=${CRAFT_PART_INSTALL}/usr/lib/jvm/java-"
+            + str(options.jlink_java_version)
+            + "-openjdk-${CRAFT_TARGET_ARCH}/"
+        )
+
+        commands.append(
+            "rm -rf ${INSTALL_ROOT} && jlink --add-modules ${deps} --output ${INSTALL_ROOT}"
+        )
+        # create /usr/bin/java link
+        commands.append(
+            "(cd ${CRAFT_PART_INSTALL} && mkdir -p usr/bin && ln -s --relative usr/lib/jvm/java-"
+            + str(options.jlink_java_version)
+            + "-openjdk-${CRAFT_TARGET_ARCH}/bin/java usr/bin/)"
+        )
+        commands.append("mkdir -p ${CRAFT_PART_INSTALL}/etc/ssl/certs/java/")
+        # link cacerts
+        commands.append(
+            "cp /etc/ssl/certs/java/cacerts ${CRAFT_PART_INSTALL}/etc/ssl/certs/java/cacerts"
```

Comment:
(Relevant reading for others coming in: https://github.com/canonical/chisel-releases/pull/308/files)

I'd rather have the docs tell the user to stage `ca-certificates-java_data`, and once we have https://github.com/canonical/craft-parts/issues/833 we can have it automatically incorporate that. What do you think?

Reply:
Hmm, we can stage ca-certificates-java_data but should avoid priming it. 
ca-certificates-java_data pulls in ca-certificates_data which is not necessary for Java keystore. 
We also need an extra step to generate Java keystore from the staged packages. 

Would not it be easier to use the build environment Java keystore ?

Reply:
We should try as much as possible not to copy things from the build environment. If the data is different on one architecture for some reason, cross-building to/from that architecture could cause issues. I'm okay with generating the Java keystore from those staged packages though.

Reply:
Actually this step is unnecessary as jlink already provides correct cacerts.